### PR TITLE
new: add bundled libelf

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area build

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cmake/modules/libbpf.cmake [hunk 1]
```diff
@@ -21,12 +21,12 @@ else()
...
         libbpf
         PREFIX "${PROJECT_BINARY_DIR}/libbpf-prefix"
-        DEPENDS zlib
+        DEPENDS zlib libelf
         URL "https://github.com/libbpf/libbpf/archive/refs/tags/v1.0.1.tar.gz"
         URL_HASH
         "SHA256=3d6afde67682c909e341bf194678a8969f17628705af25f900d5f68bd299cb03"
         CONFIGURE_COMMAND mkdir -p build root
-        BUILD_COMMAND BUILD_STATIC_ONLY=y OBJDIR=${LIBBPF_BUILD_DIR}/build DESTDIR=${LIBBPF_BUILD_DIR}/root make -C ${LIBBPF_SRC}/libbpf/src install
+        BUILD_COMMAND BUILD_STATIC_ONLY=y OBJDIR=${LIBBPF_BUILD_DIR}/build DESTDIR=${LIBBPF_BUILD_DIR}/root NO_PKG_CONFIG=1 make -e EXTRA_CFLAGS="-I${LIBELF_INCLUDE} -I${ZLIB_INCLUDE}" -e LDFLAGS="-Wl,-Bstatic" -e EXTRA_LDFLAGS="-L${LIBELF_SRC}/libelf/libelf -L${ZLIB_SRC}" -C ${LIBBPF_SRC}/libbpf/src install install_headers install_uapi_headers
```
Comment (Andreagit97) [inline cmake/modules/libbpf.cmake:29]:
```suggestion
        BUILD_COMMAND BUILD_STATIC_ONLY=y OBJDIR=${LIBBPF_BUILD_DIR}/build DESTDIR=${LIBBPF_BUILD_DIR}/root NO_PKG_CONFIG=1 make -e EXTRA_CFLAGS="-I${LIBELF_INCLUDE} -I${ZLIB_INCLUDE}" -e LDFLAGS="-Wl,-Bstatic" -e EXTRA_LDFLAGS="-L${LIBELF_SRC}/libelf/libelf -L${ZLIB_SRC}" -C ${LIBBPF_SRC}/libbpf/src install
```
We could remove these `install_headers` and `install_uapi_headers`

File: cmake/modules/libelf.cmake [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    set(LIBELF_SRC "${PROJECT_BINARY_DIR}/libelf-prefix/src")
+    set(LIBELF_INCLUDE "${LIBELF_SRC}/libelf/libelf")
+    set(LIBELF_LIB "${LIBELF_SRC}/libelf/libelf/libelf.a")
+    ExternalProject_Add(
+            libelf
+            PREFIX "${PROJECT_BINARY_DIR}/libelf-prefix"
+            DEPENDS zlib
+            URL "https://sourceware.org/elfutils/ftp/0.187/elfutils-0.187.tar.bz2"
+            #URL_HASH "SHA256=3d6afde67682c909e341bf194678a8969f17628705af25f900d5f68bd299cb03"
+            CONFIGURE_COMMAND ./configure --libdir=${LIBELF_LIB} --includedir=${LIBELF_INCLUDE} --sysconfdir=/etc --program-prefix="eu-" --enable-deterministic-archives --disable-debuginfod --disable-libdebuginfod
```
Comment (Andreagit97) [inline cmake/modules/libelf.cmake:25]: we should remove `--libdir=${LIBELF_LIB} --includedir=${LIBELF_INCLUDE}` and support the `URL_HASH`

File: userspace/libscap/CMakeLists.txt [hunk 3]
```diff
@@ -92,6 +90,7 @@ if(USE_BUNDLED_ZLIB AND NOT MINIMAL_BUILD)
 	add_dependencies(scap zlib)
 endif()
 
+
```
Comment (Andreagit97) [inline userspace/libscap/CMakeLists.txt:93]: we should remove this

**Review by Andreagit97 — COMMENTED** [review]
LGTM

**Comment by FedeDP** [comment]
Results:
* libelf is now properly built as bundled dep
* libbpf links our own libelf and our own zlib
* final Falco package does not depend upon libelf nor zlib from system
* libelf links our own zlib
* curl links our own zlib
* protobuf does not depend upon openssl and uses our own zlib
* zlib cmake is instructed to build as static lib
* pman correctly links libelf too
* multiple libscap engines need zlib
* multiple libscap engines DO NOT need zlib (cleaned up `scap-int.h` include)

**Comment by FedeDP** [comment]
TODO: 
* fix alpine Falco build
* improve libelf `CONFIGURE_COMMAND`, with something like:
```
CONFIGURE_COMMAND ./configure --libdir=${ZLIB_SRC} --includedir=${ZLIB_INCLUDE} --enable-deterministic-archives --disable-debuginfod --disable-libdebuginfod
```
That should be working but is not.

**Comment by FedeDP** [comment]
Waiting on #689 to get merged to rebase this one and squash commits :)

**Comment by FedeDP** [comment]
/hold
Trying to understand why is Falco build-musl job failing... (https://app.circleci.com/pipelines/github/falcosecurity/falco/3318/workflows/47ca1895-083d-4512-938c-e879aa1c613b/jobs/28781).

**Comment by FedeDP** [comment]
/unhold
all builds passed on Falco: https://github.com/falcosecurity/falco/pull/2277

File: cmake/modules/libbpf.cmake [hunk 4]
```diff
@@ -21,12 +21,12 @@ else()
...
         libbpf
         PREFIX "${PROJECT_BINARY_DIR}/libbpf-prefix"
-        DEPENDS zlib
+        DEPENDS zlib libelf
         URL "https://github.com/libbpf/libbpf/archive/refs/tags/v1.0.1.tar.gz"
         URL_HASH
         "SHA256=3d6afde67682c909e341bf194678a8969f17628705af25f900d5f68bd299cb03"
         CONFIGURE_COMMAND mkdir -p build root
-        BUILD_COMMAND BUILD_STATIC_ONLY=y OBJDIR=${LIBBPF_BUILD_DIR}/build DESTDIR=${LIBBPF_BUILD_DIR}/root make -C ${LIBBPF_SRC}/libbpf/src install install_uapi_headers
+        BUILD_COMMAND ${CMD_MAKE} BUILD_STATIC_ONLY=y OBJDIR=${LIBBPF_BUILD_DIR}/build DESTDIR=${LIBBPF_BUILD_DIR}/root NO_PKG_CONFIG=1 "EXTRA_CFLAGS=-I${LIBELF_INCLUDE} -I${ZLIB_INCLUDE}" "LDFLAGS=-Wl,-Bstatic" "EXTRA_LDFLAGS=-L${LIBELF_SRC}/libelf/libelf -L${ZLIB_SRC}" -C ${LIBBPF_SRC}/libbpf/src install install_uapi_headers
```
Comment (FedeDP) [inline cmake/modules/libbpf.cmake:29]: Properly link against our own built libelf and zlib. Avoid using `pkg_config` to find the correct libraries.

File: cmake/modules/libbpf.cmake [hunk 5]
```diff
@@ -21,12 +21,12 @@ else()
     ExternalProject_Add(
         libbpf
         PREFIX "${PROJECT_BINARY_DIR}/libbpf-prefix"
-        DEPENDS zlib
+        DEPENDS zlib libelf
```
Comment (FedeDP) [inline cmake/modules/libbpf.cmake:24]: It does depend upon our own libelf!

File: cmake/modules/curl.cmake [hunk 6]
```diff
@@ -19,10 +20,17 @@ else()
 		set(CURL_SSL_OPTION "--with-ssl")
 	else()
 		set(CURL_SSL_OPTION "--with-ssl=${OPENSSL_INSTALL_DIR}")
-		message(STATUS "Using bundled curl in '${CURL_BUNDLE_DIR}'")
-		message(STATUS "Using SSL for curl in '${CURL_SSL_OPTION}'")
+		message(STATUS "Using SSL for curl in '${OPENSSL_INSTALL_DIR}'")
 	endif()
 
+	if(NOT USE_BUNDLED_ZLIB)
```
Comment (FedeDP) [inline cmake/modules/curl.cmake:26]: Properly set our own zlib to build curl.

File: cmake/modules/libelf.cmake [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#
```
Comment (FedeDP) [inline cmake/modules/libelf.cmake:1]: Libelf does now support a BUNDLED mode.

File: cmake/modules/protobuf.cmake [hunk 8]
```diff
@@ -26,11 +25,11 @@ else()
 		message(STATUS "Using bundled protobuf in '${PROTOBUF_SRC}'")
 		ExternalProject_Add(protobuf
 			PREFIX "${PROJECT_BINARY_DIR}/protobuf-prefix"
-			DEPENDS openssl zlib
+			DEPENDS zlib
```
Comment (FedeDP) [inline cmake/modules/protobuf.cmake:28]: Protobuf does not seem to need openssl.

Comment (FedeDP) [reply cmake/modules/protobuf.cmake:28]: https://github.com/protocolbuffers/protobuf/blob/main/src/README.md#c-protobuf---unix

File: cmake/modules/zlib.cmake [hunk 9]
```diff
@@ -38,7 +38,7 @@ else()
 				PREFIX "${PROJECT_BINARY_DIR}/zlib-prefix"
 				URL "https://github.com/madler/zlib/archive/v1.2.11.tar.gz"
 				URL_HASH "SHA256=629380c90a77b964d896ed37163f5c3a34f6e6d897311f1df2a7016355c45eff"
-				CONFIGURE_COMMAND ./configure --prefix=${ZLIB_SRC}
+				CONFIGURE_COMMAND ./configure --prefix=${ZLIB_SRC} --static
```
Comment (FedeDP) [inline cmake/modules/zlib.cmake:41]: Force static only generation of libz.

File: userspace/libscap/CMakeLists.txt [hunk 10]
```diff
@@ -150,6 +144,11 @@ add_library(scap_event_schema
 		../../driver/event_table.c
 		../../driver/flags_table.c
 		../../driver/tp_table.c)
+
+if(USE_BUNDLED_ZLIB AND NOT MINIMAL_BUILD)
+	add_dependencies(scap_event_schema zlib)
```
Comment (FedeDP) [inline userspace/libscap/CMakeLists.txt:149]: Multiple scap deps require zlib.

File: userspace/libscap/CMakeLists.txt [hunk 11]
```diff
@@ -188,6 +191,10 @@ add_subdirectory(engine/udig)
 target_link_libraries(scap scap_engine_udig)
 
 if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+	if (NOT MINIMAL_BUILD OR BUILD_LIBSCAP_MODERN_BPF)
+		include(libelf)
```
Comment (FedeDP) [inline userspace/libscap/CMakeLists.txt:195]: Include libelf for modern and old bpf probes.

File: userspace/libscap/engine/bpf/bpf.h [hunk 12]
```diff
@@ -18,6 +18,7 @@ limitations under the License.
 #pragma once
 
 #include <stdbool.h>
+#include <limits.h>
```
Comment (FedeDP) [inline userspace/libscap/engine/bpf/bpf.h:21]: Needed by musl.

**Review by Andreagit97 — APPROVED** [review]
/approve

**Review by leogr — APPROVED** [review]
:package:

**Pull request merged**